File: packages/cli/src/commands/keys.ts
import { apiRequest } from '../lib/api.js'

interface ApiKey {
  id: string
  name: string
  prefix: string
  createdAt: string
  lastUsedAt?: string
}

interface CreateKeyResponse {
  id: string
  name: string
  key: string
}

export async function keysListAction(opts: { json?: boolean }): Promise<void> {
  const res = await apiRequest<{ keys: ApiKey[] }>('GET', '/keys')

  if (!res.ok) {
    console.error(`Error: ${res.status}`)
    if (opts.json) console.log(JSON.stringify(res.data))
    process.exit(1)
  }

  const keys = res.data.keys ?? []

  if (opts.json) {
    console.log(JSON.stringify(keys, null, 2))
    return
  }

  if (keys.length === 0) {
    console.log('No API keys found.')
    return
  }

  console.log('ID\t\tName\t\tPrefix\t\tCreated')
  console.log('─'.repeat(60))
  for (const k of keys) {
    console.log(`${k.id}\t${k.name}\t${k.prefix}...\t${k.createdAt}`)
  }
}

export async function keysCreateAction(
  opts: { name: string; json?: boolean }
): Promise<void> {
  const res = await apiRequest<CreateKeyResponse>('POST', '/keys', {
    name: opts.name,
  })

  if (!res.ok) {
    console.error(`Error: ${res.status}`)
    if (opts.json) console.log(JSON.stringify(res.data))
    process.exit(1)
  }

  if (opts.json) {
    console.log(JSON.stringify(res.data, null, 2))
  } else {
    console.log(`Key created: ${res.data.key}`)
    console.log(`(Save this key — it won't be shown again)`)
  }
}

export async function keysRevokeAction(
  keyId: string,
  opts: { json?: boolean }
): Promise<void> {
  const res = await apiRequest('DELETE', `/keys/${keyId}`)

  if (!res.ok) {
    console.error(`Error: ${res.status}`)
    if (opts.json) console.log(JSON.stringify(res.data))
    process.exit(1)
  }

  if (opts.json) {
    console.log(JSON.stringify({ status: 'ok', id: keyId }))
  } else {
    console.log(`Key ${keyId} revoked.`)
  }
}

File: packages/cli-rs/src/commands/keys.rs
use anyhow::Result;
use serde::Deserialize;

use crate::api;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct ApiKey {
    id: String,
    name: String,
    prefix: String,
    created_at: String,
    #[allow(dead_code)]
    last_used_at: Option<String>,
}

#[derive(Deserialize)]
struct KeysListResponse {
    data: Vec<ApiKey>,
}

#[derive(Deserialize)]
struct KeyCreateResponse {
    #[allow(dead_code)]
    id: String,
    #[allow(dead_code)]
    name: String,
    key: String,
}

pub async fn list_action(json: bool) -> Result<()> {
    let resp = api::api_request::<serde_json::Value>("GET", "/keys", None::<&()>).await?;

    if json {
        println!("{}", serde_json::to_string_pretty(&resp.data)?);
        if !resp.ok {
            std::process::exit(1);
        }
        return Ok(());
    }

    if !resp.ok {
        eprintln!("Error: {}", resp.status);
        std::process::exit(1);
    }

    let list: KeysListResponse = serde_json::from_value(resp.data)?;

    if list.data.is_empty() {
        println!("No API keys found.");
        return Ok(());
    }

    println!(
        "{:<16} {:<16} {:<16} {}",
        "ID", "Name", "Prefix", "Created"
    );
    println!("{}", "─".repeat(64));
    for key in &list.data {
        println!(
            "{:<16} {:<16} {:<16} {}",
            truncate(&key.id, 14),
            truncate(&key.name, 14),
            &key.prefix,
            &key.created_at,
        );
    }

    Ok(())
}

pub async fn create_action(name: &str, json: bool) -> Result<()> {
    let body = serde_json::json!({ "name": name });
    let resp = api::api_request::<serde_json::Value>("POST", "/keys", Some(&body)).await?;

    if json {
        println!("{}", serde_json::to_string_pretty(&resp.data)?);
        if !resp.ok {
            std::process::exit(1);
        }
        return Ok(());
    }

    if !resp.ok {
        eprintln!("Error: {}", resp.status);
        std::process::exit(1);
    }

    let created: KeyCreateResponse = serde_json::from_value(resp.data)?;
    println!("Key created: {}", created.key);
    println!("(Save this key — it won't be shown again)");

    Ok(())
}

pub async fn revoke_action(key_id: &str, json: bool) -> Result<()> {
    let path = format!("/keys/{}", key_id);
    let resp = api::api_request::<serde_json::Value>("DELETE", &path, None::<&()>).await?;

    if json {
        println!("{}", serde_json::to_string_pretty(&resp.data)?);
        if !resp.ok {
            std::process::exit(1);
        }
        return Ok(());
    }

    if !resp.ok {
        eprintln!("Error: {}", resp.status);
        std::process::exit(1);
    }

    println!("Key {} revoked.", key_id);

    Ok(())
}

fn truncate(s: &str, max: usize) -> String {
    if s.len() > max {
        format!("{}…", &s[..max - 1])
    } else {
        s.to_string()
    }
}
